Algorithm Improvements for Fast Lomb Scargle

kkrutkowski opened the pull request:
<!-- These comments are hidden when you submit the pull request,
so you do not need to remove them! -->

<!-- Please be sure to check out our contributing guidelines,
https://github.com/astropy/astropy/blob/main/CONTRIBUTING.md .
Please be sure to check out our code of conduct,
https://github.com/astropy/astropy/blob/main/CODE_OF_CONDUCT.md . -->

<!-- If you are new or need to be re-acquainted with Astropy
contributing workflow, please see
https://docs.astropy.org/en/latest/development/quickstart.html .
There is even a practical example at
https://docs.astropy.org/en/latest/development/git_edit_workflow_examples.html . -->

<!-- Please just have a quick search on GitHub to see if a similar
pull request has already been posted.
We have old closed pull requests that might provide useful code or ideas
that directly tie in with your pull request. -->

<!-- We have several automatic features that run when a pull request is open.
They can appear daunting but do not worry because maintainers will help
you navigate them, if necessary. -->

### Description
This pull request adds the [Low-Rank Approximation](https://arxiv.org/abs/1701.04492) implementation of the Nonuniform Fast Fourier Transform as an alternative backend to the trig_sum() utility in the Lomb-Scargle utils file and adjusts the tests and docs accordingly.

<!-- Provide a general description of what your pull request does.
Complete the following sentence and add relevant details as you see fit. -->

The new implementation is up 4x slower, when compared to existing one, but also 8 to 10 orders of magnitude more accurate, nearly matching the accuracy of the slow method.

While not having a significant impact on the typical accuracy of the standard _fast_ implementation (except eliminating the negative powers issue) it significantly improves the _fastchi2_ method, causing it to nearly match the slow _chi2_ implementation in the terms of accuracy.

~~The new algorithm is now set as default for both _fast_ and _fastchi2_ methods, although it may be better keep the original defaults in the _fast_ method to avoid the slow-down. Additionally it requires SciPy installed to be functional (for the [jv](https://docs.scipy.org/doc/scipy/reference/generated/scipy.special.jv.html) bessel functions - if `prefer_lra=True` is passed, but the SciPy is not available it falls back to the current code printing a warning suggesting explicitely passing `prefer_lra=False`~~

The behaviour got modified after preliminary review and now the Low Rank approximation gets chosen by passing `algorithm='lra'` to either of the `fast` methods or the `trig_sum` function. Additionally the Bessel function approximation relies on the Taylor series around x=0 and as such no longer relies on SciPy for evaluation.

The new algorithm is default for the `fastchi2` and `auto` methods, although the less accurate approximation is kept as the default option for the `fast` method if it is called directly. The old (default) trig-sum algorithm additionally can be specified manually by passing `algorithm='fasper'` (sharing the name of the original Fortran subroutine published by Press and Rybicki in 1989), although currently it remains the default behaviour.

<!-- In addition please ensure that the pull request title is descriptive
and allows maintainers to infer the applicable subpackage(s). -->

Technically the code almost certainly could have been made faster than the 'old' fast method if rewritten in C (as the way the operations are performed with NumPy is at least awkward and probably far from optimal), although I'm not sure if it's even worth consideration, as the SciPy-independent `jv` implementation would probably take over 1000 lines of code, with the full implementation nearing 2 thousand (and causing a significant bloat to the library as a whole).

EDIT: Using the locally converging Taylor series would probably allow to fit the whole implementation in ~ 500 LOC making the rewrite significantly more feasible

<!-- If the pull request closes any open issues you can add this.
If you replace <Issue Number> with a number, GitHub will automatically link it.
If this pull request is unrelated to any issues, please remove
the following line. -->

Fixes #5941, #9523 and #11302

<!-- Optional opt-out -->

kkrutkowski committed:
Add the Low Rank Approximation

Add the Low Rank Approximation as the alternative NUFFT algorithm of the trigsum utility

pllim commented:
There are ways to handle that doctest but maybe want to wait until the initial round of review first. Should fix the pre-commit check though. Are you able to squash your commits while you are at it? Thanks!

kkrutkowski commented:
> There are ways to handle that doctest but maybe want to wait until the initial round of review first. Should fix the pre-commit check though. Are you able to squash your commits while you are at it? Thanks!

Thanks for reply, but I'm not sure how can I squash this commit - the git/ssh interface doesn't seem to work and I have no access to the squash option in the GUI (although I may just be doing something wrong - I'll definitely try to fix that later). 

Am I supposed to just reopen this pull request from a different branch (if rebasing won't start working again soon)? Sorry once again. 

Btw I'll fix the formatting soon, although I'm not really sure how the imports should be sorted to make it work - any advices? Sorry for the chaos.

pllim commented:
For a significant change as this, it would be tough to solely rely on GH web interface. Are you not able to clone your fork locally and then work on your branch from your machine?

pllim commented:
Please have a look at our contributing guide at https://docs.astropy.org/en/latest/index_dev.html in case it helps. It also spams our CI and notification system when you push out a lot of commits via the GH web interface like this.

kkrutkowski commented:
> For a significant change as this, it would be tough to solely rely on GH web interface. Are you not able to clone your fork locally and then work on your branch from your machine?

Thanks for quick reply, cloning and working/squashing locally works just fine, but pushing/rebasing the squashed repository for some reason fails. It may have something to do with my (incorrect) system configuration, so I'll have to try to fix that.

I've also seen the contributing guide, although I'm not sure how to make a use of it. Squash rebasing is something I'll try to perform in the next few days, but I'm not sure while it fails. 

Also ruff locally doesn't give any issues when running on the same code, but in AstroPy's repo returns 3 issues (even after applying 'ruff format' on the same file just before uploading), so I have no idea what happens here.

Thanks for help and sorry for the trouble once again.

pllim commented:
After a rebase, you have to do a force push. If you are not sure it would work, a backup branch where you can recover from would be prudent. As for the pre-commit, I can try to apply autofix and see.

pllim commented:
pre-commit.ci autofix

kkrutkowski commented:
> After a rebase, you have to do a force push. If you are not sure it would work, a backup branch where you can recover from would be prudent. As for the pre-commit, I can try to apply autofix and see.

Thanks, looks like now somehow everything is just fine. Force push is where the error happens. I'll try to fix it later and/or try with `gh` instead of `git`. Or try from differenc PC, maybe this will (somehow) solve the issue.

pllim commented:
FWIW force push is pretty much mandatory if you squash commits because the PR history has changed. But do what you will for now. As long as the diff is clean here, review should be possible. Thanks again!

kkrutkowski commented:
> FWIW force push is pretty much mandatory if you squash commits because the PR history has changed. But do what you will for now. As long as the diff is clean here, review should be possible. Thanks again!

Thanks for help once again, using `gh` (github-cli) instead of `git` (or rather with, as I've used the gh only to log into my repo) directly somehow have worked I guess.  I'm leaving it here if someone had the same issue in the future.

So I guess that might be enough. Sorry for inconvenience.

neutrinoceros commented on astropy/timeseries/periodograms/lombscargle/implementations/fast_impl.py:
```diff
@@ -14,11 +14,12 @@ def lombscargle_fast(
     fit_mean=True,
     normalization="standard",
     use_fft=True,
+    prefer_lra=True,
     trig_sum_kwds=None,
```
any new argument to a public API function needs to be added at the *end* of the signature for backward compatibility.  Even if this module is deeply nested and therefore officially private, it is not clearly marked as such so, no reason to break backward compatibility when the cost of preserving it is so low. It is also good practice to avoid adding positional-allowed arguments beyond 3 or 4 args.
```suggestion
    trig_sum_kwds=None,
    *,
    prefer_lra=True,
```

neutrinoceros replied:
Additionally, I'd recommend we avoid creating another [boolean trap](https://adamj.eu/tech/2021/07/10/python-type-hints-how-to-avoid-the-boolean-trap/). Can this be expressed as `method='low-rank-approximation'` VS `method='press-rybicky'` (which would be the default) ?

neutrinoceros commented on astropy/timeseries/periodograms/lombscargle/implementations/fast_impl.py:
```diff
@@ -39,9 +40,12 @@ def lombscargle_fast(
         the result. Otherwise, use a slower O[N^2] algorithm
+    prefer_lra : bool
+        if True and SciPy is available, use the more accurate Low Rank Approximation by Ruiz-Antolin and Townsend.
+        If False or SciPy is not available use Press & Rybicki's Lagrangian extirpolation instead.
```
I would consider it surprising that I, as a user, explicitly request LRA be used and `astropy` silently ignoring it on the ground that I don't have `scipy` installed. I would recommend making it a hard error instead.

neutrinoceros commented on astropy/timeseries/periodograms/lombscargle/implementations/fastchi2_impl.py:
```diff
@@ -43,6 +44,9 @@ def lombscargle_fastchi2(
         if True, pre-center the data by subtracting the weighted mean
         of the input data. This is especially important if ``fit_mean = False``
+    prefer_lra : bool
+        if True and SciPy is available, use the more accurate Low Rank Approximation by Ruiz-Antolin and Townsend.
```
same here

neutrinoceros commented on astropy/timeseries/periodograms/lombscargle/implementations/utils.py:
```diff
@@ -1,7 +1,15 @@
+from astropy.utils.exceptions import AstropyUserWarning
+
+if HAS_SCIPY:
+    from scipy.fft import next_fast_len
+    from scipy.special import jv, lambertw
```
I think we should avoid importing from `scipy` at the top level of any module that doesn't have a hard dependency on it: these imports are costly so we want to run them only as needed

kkrutkowski replied:
> I think we should avoid importing from `scipy` at the top level of any module that doesn't have a hard dependency on it: these imports are costly so we want to run them only as needed

One more question - would you still suggest splitting the utils file now, when the SciPy imports are gone? (Btw the locally converging series somehow ended up noticeably faster, than the SciPy's implementation, which I totally didn't expect).

If so how would you split the current 'utils' into different modules? The only idea that seems to make sense to me now would be changing utils into a directory, moving mle.py and utils.py (renamed to trig_sum.py?) and creating a separate (sub?)-module for each nufft method.

Passing the nufft variant as 'algorithm' instead of 'method' also seems to be a little bit clearer, as the 'method' could get confused with the methods of the main lomb-scargle class, but I'm not sure if that's the right way to approach the problem.

Sorry for asking that many questions and thanks for comments.

neutrinoceros commented on astropy/timeseries/periodograms/lombscargle/implementations/utils.py:
```diff
@@ -79,7 +87,213 @@ def extirpolate(x, y, N=None, M=4):
 
 
-def trig_sum(t, h, df, N, f0=0, freq_factor=1, oversampling=5, use_fft=True, Mfft=4):
+if HAS_SCIPY:
```
If you want to define APIs that are only available if scipy is installed, I think it would make more sense to put them into a dedicated module instead of conditionally skipping their definition.

neutrinoceros reviewed (dismissed):
This is a high-level review in that I didn't scrutinize with the details of the implementation, but focused on the way the APIs are evolved. I wrote a couple targeted comments that actually apply in more than one place. In particular, I recommend avoiding the addition of a boolean `prefer_lra` parameter, that should instead be something like `method='lra'`.

kkrutkowski commented:
> This is a high-level review in that I didn't scrutinize with the details of the implementation, but focused on the way the APIs are evolved. I wrote a couple targeted comments that actually apply in more than one place. In particular, I recommend avoiding the addition of a boolean `prefer_lra` parameter, that should instead be something like `method='lra'`.

Thanks for reply, that definitely makes more sense from the users' perspective. Although I have one more question - how do you see it implemented for the fastchi2 method (which is the one affected significantly more, when compared to the standard 'fast' by that update?

The way I see it using the name 'fasper' as an alias for current 'fast' method (named after the original Fortran subroutine with the same name published by Press and Rybicki - that name to me looks more concise and seems to be used more often, than explicitely refering to this by their names, but that may be just me not knowing any representative sample) and making 'method=lra' an alias for the new trig_sum variant. Would something like that be fine?

Also as for the separate submodule - are you suggesting creating a separate file with an alternative definition of the trig_sum and conditionally importing it from implementations/utils if lra is selected as the trig_sum() approximation method?

I don't really see a reason for repeating all of the fast/fastchi2 methods code, as essentially it's still the same method by their high-level details - they differ only in their 'extirpolation', and as such at least ~ ¾ of existing code can be reused.

Also I assume, that the SciPy-independent implementation of the jv() function might get added in the future, but you're definitely right that I shouldn't threat the SciPy import as only a temporary solution here.

neutrinoceros commented:
> how do you see it implemented for the fastchi2 method (which is the one affected significantly more, when compared to the standard 'fast' by that update?

I'm not sure I'm familiar enough with the details to answer this question right now. Let me clarify that when I proposed `method` as a argument in place of `prefer_lra`, I was taking a guess; if this argument name isn't appropriate, I trust you can come up with a better one. The important part of my point is that we should avoid boolean arguments everywhere that more than 2 options *may* exist (even if there are exactly 2 at the time of writing).

>The way I see it using the name 'fasper' as an alias for current 'fast' method (named after the original Fortran subroutine with the same name published by Press and Rybicki - that name to me looks more concise and seems to be used more often, than explicitely refering to this by their names, but that may be just me not knowing any representative sample) and making 'method=lra' an alias for the new trig_sum variant. Would something like that be fine?

I don't think we should always favor conciseness over clarity. I can be argued that conciseness is fine for domain-specific jargon though; if you think it's reasonable to expect that any potential user would immediately understand what `fasper` means, then I don't have any objections. Otherwise, I recommend being as explicit as possible.

>Also as for the separate submodule - are you suggesting creating a separate file with an alternative definition of the trig_sum and conditionally importing it from implementations/utils if lra is selected as the trig_sum() approximation method?

no. I'm suggesting to move conditional definitions to a separate module that would have a hard-requirement on scipy, but only be imported from conditionally.

>I don't really see a reason for repeating all of the fast/fastchi2 methods code, as essentially it's still the same method by their high-level details - they differ only in their 'extirpolation', and as such at least ~ ¾ of existing code can be reused.

by all means, *do* reuse existing code whenever you can. I certainly didn't mean to suggest otherwise. Is there anything I need to clarify from my initial reply ?

kkrutkowski commented:
> > how do you see it implemented for the fastchi2 method (which is the one affected significantly more, when compared to the standard 'fast' by that update?
> 
> I'm not sure I'm familiar enough with the details to answer this question right now. Let me clarify that when I proposed `method` as a argument in place of `prefer_lra`, I was taking a guess; if this argument name isn't appropriate, I trust you can come up with a better one. The important part of my point is that we should avoid boolean arguments everywhere that more than 2 options *may* exist (even if there are exactly 2 at the time of writing).
> 
> >The way I see it using the name 'fasper' as an alias for current 'fast' method (named after the original Fortran subroutine with the same name published by Press and Rybicki - that name to me looks more concise and seems to be used more often, than explicitely refering to this by their names, but that may be just me not knowing any representative sample) and making 'method=lra' an alias for the new trig_sum variant. Would something like that be fine?
> 
> I don't think we should always favor conciseness over clarity. I can be argued that conciseness is fine for domain-specific jargon though; if you think it's reasonable to expect that any potential user would immediately understand what `fasper` means, then I don't have any objections. Otherwise, I recommend being as explicit as possible.
> 
> >Also as for the separate submodule - are you suggesting creating a separate file with an alternative definition of the trig_sum and conditionally importing it from implementations/utils if lra is selected as the trig_sum() approximation method?
> 
> no. I'm suggesting to move conditional definitions to a separate module that would have a hard-requirement on scipy, but only be imported from conditionally.
> 
> >I don't really see a reason for repeating all of the fast/fastchi2 methods code, as essentially it's still the same method by their high-level details - they differ only in their 'extirpolation', and as such at least ~ ¾ of existing code can be reused.
> 
> by all means, *do* reuse existing code whenever you can. I certainly didn't mean to suggest otherwise. Is there anything I need to clarify from my initial reply ?

Thanks, everything seems clear now. Also I could implement a simple fallback NumPy jv() series using gamma (and possibly factorial) from math - I'll check if it is accurate enough (and almost definitely it would be significantly slower, than the SciPy implementation), although it can be done in like ~ 15 lines of code with no external dependencies. If you thing something like that would be more suitable, than the current fallback to the much less accurate (but faster) method it might be worth consideration.

neutrinoceros commented:
On the conditions that:
- default behavior is unchanged
- explicit user requests are never ignored (raise an exception when they cannot be satisfied)
I don't see any problem with having more options available

Commit:
Remove SciPy dependency

neutrinoceros commented on astropy/timeseries/periodograms/lombscargle/implementations/fast_impl.py:
```diff
@@ -41,8 +42,15 @@ def lombscargle_fast(
+        Specify the approximation used to approximate the NUDFT of type 1. If the value is not valid falls back to the default option.
+        Currently there are two available options:
+
+        - 'fasper': use Press & Rybicki's Lagrangian extirpolation instead. This is the default option.
```
```suggestion
        - 'fasper': use Press & Rybicki's Lagrangian extirpolation. This is the default option.
```

neutrinoceros commented on astropy/timeseries/periodograms/lombscargle/implementations/fast_impl.py:
```diff
@@ -41,8 +42,15 @@ def lombscargle_fast(
+    algorithm : str, optional
+        Referenced only if use_fft is true.
+        Specify the approximation used to approximate the NUDFT of type 1. If the value is not valid falls back to the default option.
+        Currently there are two available options:
```
let's rephrase this in a future-compatible way so it doesn't risk going out of sync
```suggestion
        Supported options are:
```

neutrinoceros commented on astropy/timeseries/periodograms/lombscargle/implementations/fastchi2_impl.py:
```diff
@@ -45,6 +46,13 @@ def lombscargle_fastchi2(
     nterms : int, optional
         Number of Fourier terms in the fit
+    algorithm : str, optional
+        Referenced only if use_fft is true.
```
Is this what you mean ?
```suggestion
        This option is ignored if if use_fft is False.
```

neutrinoceros commented on astropy/timeseries/periodograms/lombscargle/implementations/fastchi2_impl.py:
```diff
@@ -45,6 +46,13 @@ def lombscargle_fastchi2(
+        Specify the approximation used to approximate the NUDFT of type 1. If the value is not valid falls back to the default option.
+        Currently there are two available options:
+
+        - 'fasper': use Press & Rybicki's Lagrangian extirpolation instead. This is the default option.
```
```suggestion
        - 'fasper': use Press & Rybicki's Lagrangian extirpolation. This is the default option.
```

neutrinoceros commented on astropy/timeseries/periodograms/lombscargle/implementations/fastchi2_impl.py:
```diff
@@ -18,6 +18,7 @@ def lombscargle_fastchi2(
     nterms=1,
     use_fft=True,
     trig_sum_kwds=None,
+    algorithm='lra',
```
```suggestion
    *,
    algorithm="lra",
```
(note that I'm using double quotes for strings because that's what our formatter, ruff format, will enforce anyway)

neutrinoceros commented on astropy/timeseries/periodograms/lombscargle/implementations/tests/test_utils.py:
```diff
@@ -64,6 +64,7 @@ def trig_sum_data():
 @pytest.mark.parametrize("adjust_t", [True, False])
 @pytest.mark.parametrize("freq_factor", [1, 2])
 @pytest.mark.parametrize("df", [0.1])
+
```
```suggestion
```

neutrinoceros commented on astropy/timeseries/periodograms/lombscargle/implementations/tests/test_utils.py:
```diff
@@ -73,12 +74,22 @@ def test_trig_sum(f0, adjust_t, freq_factor, df, trig_sum_data):
         df,
         N=1000,
-        use_fft=True,
+        use_fft=False,
```
hum, maybe define a `S0, C0` pair instead of changing the meaning of `S1, C1` ?

neutrinoceros commented on astropy/timeseries/periodograms/lombscargle/tests/test_lombscargle.py:
```diff
@@ -186,10 +186,11 @@ def test_nterms_methods(
         P_expected = ls.power(frequency)
 
-        # don't use fast fft approximations here
+        # don't use fast fft approximations here if the SciPy is not available
```
looks like this comment is out of sync now
```suggestion
        # don't use fast fft approximations here
```

neutrinoceros commented on astropy/timeseries/periodograms/lombscargle/tests/test_lombscargle.py:
```diff
@@ -186,10 +186,11 @@ def test_nterms_methods(
         kwds = {}
         if "fast" in method:
-            kwds["method_kwds"] = dict(use_fft=False)
+            kwds["method_kwds"] = dict(algorithm='fasper')
```
isn't this a change in behavior ?

kkrutkowski replied:
It is and it shouldn't have passed the tests (?), as 'fasper' generaters errors of the order of 1e-3 or sometimes even significantly higher (if not all of the artifacts caused by the Runge's phenomenon go out of the grid's frequency range).

Thanks for catching, but it looks a little bit concerning, as it somehow did pass the tests. I'll delve a little bit deeper into this soon, but for now I don't see any issues that could've caused the tests to pass (while they definitely should have failed here)

neutrinoceros replied:
It's absolutely possible that we're missing a test here. Parameter space is subject to combinatoric explosion and it's often impossible that we test every supported case, especially in APIs that have many parameters (which is a good reason to avoid them whenever possible, but this ship is long gone). It'd be worth adding a test case here *before* fixing it, so we can easily make sure that the test catches a new problem no other test does, and then we might even want to merge this test as it's own PR.

kkrutkowski replied:
I've checked the code and looks like the code doesn't check if the results are not *exactly* the same for a different methods, only if they were close enough.

What I've done previously essentially made 'fasper' compare to itself, so the difference was exactly 0 - the ```
assert_allclose()
``` of course returned true in that case.

We could add ```
assert not np.array_equal()
``` to the assert_allclose() to the tests, to make sure, that the test case is not ignored (which I have done twice without noticing during last commit) - and this should fix the issue.

I'll probably do that later today - as for now the tests seem to be checking only, if 'different' methods are close enough, but they're not checking the method dispatch in any way.

kkrutkowski replied:
Imho the checking the absolute difference may not be strict enough, as for example it would ignore NaNs or if the FPU of a given system.

Also I think IEEE 754 allows for 0.5 ULP error for floating point addition/subtraction (I may be wrong here), so computing the absolute difference could technically invoke undefined behavior (although technically it should work at least almost always).

If wanting to avoid explicit negation I think checking the bitwise equality with something like ```
assert np.bitwise_xor(a, b).any()
``` seems to match the desired test's logic more closely, but I'm not sure that would be the right choice here.

kkrutkowski replied:
Imho the checking the absolute difference may not be strict enough, as for example it would ignore NaNs or if the FPU of a given system.

Also I think IEEE 754 allows for 0.5 ULP error for floating point addition/subtraction (I may be wrong here), so computing the absolute difference could technically invoke undefined behavior (although technically it should work at least almost always).

If wanting to avoid explicit negation I think checking the bitwise equality with something like ```
assert np.bitwise_xor(a, b).any()
``` seems to match the desired test's logic more closely, but I'm not sure that would be the right choice here.

neutrinoceros reviewed (commented):
CI appears to be in a much better shape now, here's another round of review !
Please pay attention to reports from `pre-commit.ci` (I cannot recommend enough going through https://docs.astropy.org/en/latest/development/quickstart.html#install-pre-commit). It's not a big deal to defer lints to until the PR is stabilized, but it seems to be the case now (well done !).

neutrinoceros commented:
> the pre-commit indeed helps and unlike the direct 'ruff' seems to be consistent with AstroPy checks in the repo.

`ruff`'s exact behavior varies depending on the exact version you have installed. Pre-commit helps ensure you're running the exact same version as CI :)

Thank you for addressing my comments. I'll try to take another and possibly deeper look later today.

Commit:
Add dispatch tests for the 'fast' method

Commit:
Add dispatch tests for the 'fast' method

neutrinoceros commented on astropy/timeseries/periodograms/lombscargle/implementations/tests/test_utils.py:
```diff
@@ -68,25 +68,42 @@ def test_trig_sum(f0, adjust_t, freq_factor, df, trig_sum_data):
-        freq_factor=freq_factor,
-        oversampling=10,
-    )
-    S2, C2 = trig_sum(
```
it looks to me like this case has been dropped in your refactor: I don't see this combo in `fast_methods[1][0]`. Is this intentional ? I think your PR should be focused on adding tests, not removing existing ones (even if there are redundancies)

kkrutkowski replied:
Yeah, that one was intended, but I may restore the original behaviour in the next commit. The main issue with it was that the combination fitting a 0-th degree trigonometric polynomial effectively just fits the mean (== 0'th cosine term).

As such no FFT is computed at all and the test literally does nothing (returning an array of constants, independent from the chosen algorithm).

As such it is supposed to fail the bitwise_xor (which the rest of options is expected to pass).
I may either conditionally disable the xor check (which I probably will) instead of just removing it, but effectively it's not checking anything - it's not as much about redundancy, but rather an 'empty' branch it is supposed to check.

neutrinoceros replied:
I see. It is puzzling that the test exist in this form. Do you think it's possible to modify it as a separate PR ?

kkrutkowski replied:
> I see. It is puzzling that the test exist in this form. Do you think it's possible to modify it as a separate PR ?

I think that definitely would cause some merge conflicts, which could be not trivial to resolve.

If anything probably restoring the old logic (while keeping the new loop here) would be optimal (and possibly later removing the 'empty' branch in a separate PR). 

Modifying the tests before pulling this one would probably be significantly more complicated.

neutrinoceros commented on astropy/timeseries/periodograms/lombscargle/implementations/tests/test_utils.py:
```diff
@@ -68,25 +68,42 @@ def test_trig_sum(f0, adjust_t, freq_factor, df, trig_sum_data):
+        assert_allclose(C0, C, atol=tol)
+
+        for prev_S, prev_C in results:
+            assert np.bitwise_xor(
+                np.ascontiguousarray(prev_S).view(np.uint8),
+                np.ascontiguousarray(S).view(np.uint8),
+            ).any(), "Output arrays are the same - dispatch test has failed"
```
This is a bit hard for me to read. I feel like the test could be much more readable if the comparison logic was encapsulated in a helper function. Maybe something like `assert_not_strictly_equal` (assuming I'm parsing this correctly), together with a helpful docstring ?

neutrinoceros replied:
Just noticed you also use this pattern in a different module. Then the helper function should live in a private `_testing.py` module in the "most recent common ancestor" directory, which I think would be `astropy/timeseries/periodograms/lombscargle/_testing.py` ?

neutrinoceros commented on astropy/timeseries/periodograms/lombscargle/implementations/utils.py:
```diff
@@ -12,6 +12,29 @@ def bitceil(N):
 
+def next_fast_len(N):
+    """
+    Find the next number greater, than N, which can be expressed
```
```suggestion
    Find the next number, greater than N, which can be expressed
```

neutrinoceros commented on astropy/timeseries/periodograms/lombscargle/implementations/utils.py:
```diff
@@ -12,6 +12,29 @@ def bitceil(N):
+def next_fast_len(N):
+    """
+    Find the next number greater, than N, which can be expressed
+    as 2^n * 3^m for integer values of n and m and m <= 4.
```
```suggestion
    as 2^n * 3^m for integer values of n and m, with m <= 4.
```

neutrinoceros commented on astropy/timeseries/periodograms/lombscargle/implementations/utils.py:
```diff
@@ -12,6 +12,29 @@ def bitceil(N):
     return 1 << int(N - 1).bit_length()
 
 
+def next_fast_len(N):
```
this function could use a couple doctests, I think

neutrinoceros commented on astropy/timeseries/periodograms/lombscargle/implementations/utils.py:
```diff
@@ -12,6 +12,29 @@ def bitceil(N):
+        Nfft //= 32
+        Nfft *= 27
+
+    return Nfft
```
I'm confused. Is it expected that this function returns `bitceil(N)` if none of the conditions are met ? This doesn't seem to be implied by the docstring. I could see this resolved by either updating the docstring, or changing the code, but really I don't presume of your intention here.

neutrinoceros replied:
> Also I've just realised, that if the grid begins at zero the np.fft.ifft could be replaced with np.fft.rfft (in the 'fasper' branch only) making the execution noticeably faster and keeping the method mathematically equivalent to the current version.

I would recommend trying this out as a follow up PR so as to make this one easier to review

neutrinoceros commented on astropy/timeseries/periodograms/lombscargle/implementations/utils.py:
```diff
@@ -79,7 +102,273 @@ def extirpolate(x, y, N=None, M=4):
+    for i in range(1, n_max + 1):
+        fact[i] = fact[i - 1] * i
+
+    result = np.zeros(np.broadcast(n, x).shape, dtype=np.float64)
```
```suggestion
    result = np.zeros(np.broadcast_shapes(n, x), dtype=np.float64)
```

neutrinoceros commented on astropy/timeseries/periodograms/lombscargle/implementations/utils.py:
```diff
@@ -79,7 +102,273 @@ def extirpolate(x, y, N=None, M=4):
+
+    # Keep the array longer to approximately ignore the factorials of negative integers in the denominator
+    n_max = int(np.max(n)) + num_terms + 10
+    fact = np.empty(n_max + 1, dtype=np.float64)
+    fact[0] = 1.0
+    for i in range(1, n_max + 1):
+        fact[i] = fact[i - 1] * i
```
I can't help feeling that there should be a more idiomatic to write this in Python, but I cannot find one that feels as readable. Oh well.

neutrinoceros commented on astropy/timeseries/periodograms/lombscargle/implementations/utils.py:
```diff
@@ -79,7 +102,273 @@ def extirpolate(x, y, N=None, M=4):
+    term_denom = fact[n]
+
+    for k in range(num_terms):
+        term_denom = fact[k] * fact[(n + k)]
```
```suggestion
        term_denom = fact[k] * fact[n + k]
```

neutrinoceros commented on astropy/timeseries/periodograms/lombscargle/implementations/utils.py:
```diff
@@ -79,7 +102,273 @@ def extirpolate(x, y, N=None, M=4):
+        }
+
+        # Find the smallest K in the LUT where eps is less than or equal to the LUT value
+        K = 16  # Default to K=16 if eps is too small
```
this assignment could be move to an `else` branch on the `for` loop
https://docs.python.org/3/tutorial/controlflow.html#else-clauses-on-loops

neutrinoceros commented on astropy/timeseries/periodograms/lombscargle/implementations/utils.py:
```diff
@@ -79,7 +102,273 @@ def extirpolate(x, y, N=None, M=4):
+
+        # Find the smallest K in the LUT where eps is less than or equal to the LUT value
+        K = 16  # Default to K=16 if eps is too small
+        for i in sorted(lut.keys()):
+            if eps >= lut[i]:
+                K = i
+                break
```
I would maybe add a comment to `lut` saying that it should remain sorted, but I don't see the point of dynamically sorting it if we control the declaration, immediately before we retrieve keys
```suggestion
        for i, threshold in lut.items():
            if eps >= threshold:
                K = i
                break
```

neutrinoceros commented on astropy/timeseries/periodograms/lombscargle/implementations/utils.py:
```diff
@@ -79,7 +102,273 @@ def extirpolate(x, y, N=None, M=4):
+    # For V: compute Chebyshev polynomials on the frequency variable.
+    omega = np.arange(N)
+    X = 2.0 * omega / N - 1
+    Tcheb_V = chebyshev_polynomials(K - 1, X)
+    V = Tcheb_V.astype(np.complex128)
```
```suggestion
    V = chebyshev_polynomials(K - 1, X).astype(np.complex128)
```

neutrinoceros reviewed (commented):
I'm not done yet, but I think it's worth giving you my comments now before it becomes overwhelming. Thanks for your patience

kkrutkowski commented on astropy/timeseries/periodograms/lombscargle/implementations/main.py:
```diff
@@ -216,8 +216,7 @@ def lombscargle(
         )
-        kwds.update(f0=f0, df=df, Nf=Nf)
-
+        kwds.update(f0=f0, df=df, Nf=Nf, algorithm="lra")
```
This one is supposed to set the algorithm to `"lra"` for the standard fast method (which defaults to `"fasper"` when called directly).

I'm not sure it that's the best way, however it keeps the 'auto' method nearly deterministic, potentially solving some of the issues. 

However if 'fast' is chosen directly it usually suggests that someone probably needs the speed (for which Python implementation is suboptimal, however the LRA is still at least ~3x slower on most hardware).

For the FastChi2 case looks a little bit different, as the systems of equations start dominating the FFT costs pretty easily (and it's significantly more sensitive to relatively minor errors, than the standard 'fast'), so here LRA being always the default makes more sense to me.

If that's a bad idea to default to LRA for 'fast' in 'auto', but keep the direct 'fast' defaults unmodified I can change it, then I would wholeheartedly agree with this suggestion.

neutrinoceros replied:
It may make perfect sense in a vacuum, but I'm less sure about changing default behavior.
I think it may be acceptable to change how `method='auto'` and/or `method='fast'` behave, but it'd be best to require inputs from `timeseries` maintainers.
ping @astrofrog

hamogu replied:
Given that we do have real user bug reports about the default method being obviously wrong in some cases (which are rightfully listed as to be closed with this PR), we absolutely think we should change the default at least for the "auto" and "fast" to be more correct, even if they are a bit slower. 
Most (not software developing) astronomers don't report errors to us at all. Either they just trust astropy to be always right or they that "goes it doesn't work, I'll stick with IDL". The fact that there are several reports about LS means that it's a significant problem that this PR addresses. Since most people use the default, we should absolutely make sure that thy get the right (or at least best) answers with the default if we can do that.

Commit:
Apply most of the suggested changes

neutrinoceros commented:
I think at this point I've done everything in my power as a generalist maintainer to sheperd this toward success, but we're reaching a point were expertise is needed in the review process. In particular, I'd like to get @astrofrog's opinion on https://github.com/astropy/astropy/pull/17842#discussion_r1988846371
(I already pinged him on the thread, but I figure a little redundancy doesn't hurt given how hard this is becoming to navigate)
Many thanks to @kkrutkowski for your patience and dedication !

kkrutkowski committed:
Re-commit removed broadcasting fix

neutrinoceros commented:
@kkrutkowski I'm sorry the bot got to your PR before another maintainer did. I can reopen, keep the bot away, and try again to get this the attention it needs. Just say the word :)

neutrinoceros commented:
I probably asked already, and I realize I may be out of favors, but could you rebase your branch on top of current `main` ? it should fix unrelated errors in doc builds

kkrutkowski committed:
Change default LS fast algorithm from 'fasper' to 'lra'

kkrutkowski committed:
Remove redundant algorithm update in Lomb-Scargle implementation

hamogu commented:
We've waited two more weeks for give people a chance and got no more comments.

@neutrinoceros You asked for changes in review. If you get a chance it would be great if you can check that your requested changes are done.

I will have a look over the weekend to see if the new commits look good to me and then we can hopefully merge soon without waiting for anyone else any longer.

kkrutkowski committed:
Add suggested note explaining differences between implementations

Added notes on the accuracy and speed trade-offs of different Lomb-Scargle implementations.

kkrutkowski commented on docs/timeseries/lombscargle.rst:
```diff
@@ -9,16 +9,17 @@ used statistical tool designed to detect periodic signals in unevenly spaced
-The code here is adapted from the `astroml`_ package ([4]_, [5]_) and the
-`gatspy`_ package ([6]_, [7]_).  For a detailed practical discussion of the
+The code here is adapted from the `astroml`_ package ([4]_, [5]_), the
+`gatspy`_ package ([6]_, [7]_) and the `FastTransforms.jl`_ library. For a detailed practical discussion of the
```
Already done I think (?). Sorry for the late edit, but I have totally forgotten about this one.
I am also not sure how much information should I put in that part - generally the LRA converges to the true NuDFT as the order of the approximation diverges to infinity (with the upper error bounds derived by the original authors), while PR89/fasper converges only as the sampling rate goes to infinity (which is not a parameter that might be changed).

**EDIT**: Also doctest now fails after the update, I'll fix that in a moment;

/home/docs/checkouts/readthedocs.org/user_builds/astropy/checkouts/17842/docs/timeseries/lombscargle.rst:19: ERROR: Unknown target name: "astroml". [docutils]
/home/docs/checkouts/readthedocs.org/user_builds/astropy/checkouts/17842/docs/timeseries/lombscargle.rst:19: ERROR: Unknown target name: "gatspy". [docutils]
/home/docs/checkouts/readthedocs.org/user_builds/astropy/checkouts/17842/docs/timeseries/lombscargle.rst:19: ERROR: Unknown target name: "fasttransforms.jl". [docutils]

**EDIT2**; Fixed

kkrutkowski committed:
Fix the formatting issues

kkrutkowski committed:
Attempt to fix the doctest

neutrinoceros commented:
I did a quick once-over and resolved most hanging threads (all mine). Here's what I identified:

This suggestion wasn't addressed:
https://github.com/astropy/astropy/pull/17842#discussion_r1988897407
I won't block over it, feel free to close this thread in any case.

There were a couple threads where I raised concerns about changes in default behavior, for instance
https://github.com/astropy/astropy/pull/17842#discussion_r1988846371
I'm not resolving them for visibility, but I do agree with @hamogu that it's okay to change the default if it isn't always correct, given that the release that'll land this patch is going to be major any way. I would however prefer to have it clearly mentioned on the changelog and in docstrings (using `.. versionchanged: 8.0` directives).

One possible follow up was discussed at
https://github.com/astropy/astropy/pull/17842#discussion_r1988901655
if we don't want to loose sight of it, we should make it into a new issue. I'll let you guys be the judges of that.

astrofrog commented:
With my timeseries maintainer hat on, I'm now taking a look at this (apologies for the long delay). I'll post some review comments, and @kkrutkowski if you end up not having time to address these, I'll see if I can address them myself and make sure we merge this in time for 8.0.0

astrofrog commented on astropy/timeseries/periodograms/lombscargle/implementations/utils.py:
```diff
@@ -107,50 +394,81 @@ def trig_sum(t, h, df, N, f0=0, freq_factor=1, oversampling=5, use_fft=True, Mff
+
+            fftgrid = nufft1_lra(tnorm, h, N, eps)
+
+            if t0 > 0:
```
Should this be t0 != 0? (like on line 465). Currently I think this will not work properly when negative values are present:

```python
In [1]: import numpy as np
   ...: from astropy.timeseries.periodograms.lombscargle.implementations.utils import trig_sum

In [2]: rng = np.random.default_rng(0)
   ...: N = 30
   ...: t = np.sort(rng.uniform(-100, 100, N))
   ...: h = rng.standard_normal(N)
   ...: df, Nf, f0 = 0.01, 50, 0.05

In [3]: S_true, C_true = trig_sum(t, h, df, N=Nf, f0=f0, use_fft=False)
   ...: S_lra,  C_lra  = trig_sum(t, h, df, N=Nf, f0=f0, use_fft=True, algorithm='lra')

In [4]: # Shift so t.min() > 0 (same data, just translated)                                   
   ...: t_pos = t - t.min() + 1.0
   ...: S_true2, C_true2 = trig_sum(t_pos, h, df, N=Nf, f0=f0, use_fft=False)
   ...: S_lra2,  C_lra2  = trig_sum(t_pos, h, df, N=Nf, f0=f0, use_fft=True, algorithm='lra')

In [5]:   print("t.min() < 0: max|lra - slow| C, S =",
   ...:         np.max(np.abs(C_lra - C_true)),  np.max(np.abs(S_lra - S_true)))
   ...:   print("t.min() > 0: max|lra - slow| C, S =",
   ...:         np.max(np.abs(C_lra2 - C_true2)), np.max(np.abs(S_lra2 - S_true2)))
t.min() < 0: max|lra - slow| C, S = 10.830098000379738 13.39817781900997
t.min() > 0: max|lra - slow| C, S = 1.5694112676101213e-12 9.410250356722827e-13
```

kkrutkowski replied:
> Should this be t0 != 0? (like on line 465). Currently I think this will not work properly when negative values are present:
> 
> ```python
> In [1]: import numpy as np
>    ...: from astropy.timeseries.periodograms.lombscargle.implementations.utils import trig_sum
> 
> In [2]: rng = np.random.default_rng(0)
>    ...: N = 30
>    ...: t = np.sort(rng.uniform(-100, 100, N))
>    ...: h = rng.standard_normal(N)
>    ...: df, Nf, f0 = 0.01, 50, 0.05
> 
> In [3]: S_true, C_true = trig_sum(t, h, df, N=Nf, f0=f0, use_fft=False)
>    ...: S_lra,  C_lra  = trig_sum(t, h, df, N=Nf, f0=f0, use_fft=True, algorithm='lra')
> 
> In [4]: # Shift so t.min() > 0 (same data, just translated)                                   
>    ...: t_pos = t - t.min() + 1.0
>    ...: S_true2, C_true2 = trig_sum(t_pos, h, df, N=Nf, f0=f0, use_fft=False)
>    ...: S_lra2,  C_lra2  = trig_sum(t_pos, h, df, N=Nf, f0=f0, use_fft=True, algorithm='lra')
> 
> In [5]:   print("t.min() < 0: max|lra - slow| C, S =",
>    ...:         np.max(np.abs(C_lra - C_true)),  np.max(np.abs(S_lra - S_true)))
>    ...:   print("t.min() > 0: max|lra - slow| C, S =",
>    ...:         np.max(np.abs(C_lra2 - C_true2)), np.max(np.abs(S_lra2 - S_true2)))
> t.min() < 0: max|lra - slow| C, S = 10.830098000379738 13.39817781900997
> t.min() > 0: max|lra - slow| C, S = 1.5694112676101213e-12 9.410250356722827e-13
> ```

Thanks for noticing the issue, I've totally forgotten about this one. Validation step prohibits the algorithm to be 'fed' negative frequencies and for some reason I assumed it's exactly the same for times, but now I see that it was not the case. I'll change this one soon as well.

astrofrog commented on astropy/timeseries/periodograms/lombscargle/implementations/utils.py:
```diff
@@ -79,7 +102,271 @@ def extirpolate(x, y, N=None, M=4):
+            16: 4.6e-16,
+        }
+
+        # Find the smallest K in the LUT where eps is less than or equal to the LUT value
```
Should it be 'greater than' not 'less than'?

astrofrog commented on astropy/timeseries/periodograms/lombscargle/implementations/utils.py:
```diff
@@ -79,7 +102,271 @@ def extirpolate(x, y, N=None, M=4):
+    N : int
+        Number of desired frequency bins.
+    eps : float
+        Desired relative error tolerance.
```
Isn't this used as an absolute tolerance?

kkrutkowski replied:
This one is used as the expected value of average relative error of the NuFFT, which is proportional to the absolute error of the periodogram if the normalization is set to either "standard" or "model". 

I think I should adjust it given the changes from original code using Lambert W, because now it's not really a 'tolerance' anymore, but here (within NuFFT) it still is relative.

astrofrog commented on astropy/timeseries/periodograms/lombscargle/implementations/utils.py:
```diff
@@ -79,7 +102,271 @@ def extirpolate(x, y, N=None, M=4):
 
 
-def trig_sum(t, h, df, N, f0=0, freq_factor=1, oversampling=5, use_fft=True, Mfft=4):
+def jn(n, x, num_terms=10):
```
I think it would be good to add a test (which would work when scipy is installed) which compares this implementation to scipy. Currently there are some values for which the two disagree a lot, though not clear if it would happen in real-life situations:

```
In [7]: from astropy.timeseries.periodograms.lombscargle.implementations.utils import jn

In [8]: from scipy.special import jv

In [10]: for n in [5, 1, -3, -5, -7]:
    ...:     x = 0.01
    ...:     print(f"jn({n},{x}) = {jn(n, x):12.3e}   jv({n},{x}) = {jv(n, x):12.3e}")
    ...: 

jn(5,0.01) =    2.604e-14   jv(5,0.01) =    2.604e-14
jn(1,0.01) =    5.000e-03   jv(1,0.01) =    5.000e-03
jn(-3,0.01) =    6.097e-06   jv(-3,0.01) =   -2.083e-08
jn(-5,0.01) =    8.017e+03   jv(-5,0.01) =   -2.604e-14
jn(-7,0.01) =    2.540e+12   jv(-7,0.01) =   -1.550e-20
```

I'll suggest some edits below that result in better agreement

astrofrog commented on astropy/timeseries/periodograms/lombscargle/implementations/utils.py:
```diff
@@ -79,7 +102,271 @@ def extirpolate(x, y, N=None, M=4):
+        result += term_numer / term_denom
+        term_numer *= -half_x2
+
+    return result
```
```suggestion
    # Use J_{-m}(x) = (-1)^m J_m(x) to reduce to non-negative order
    m = np.abs(n)
    sign = np.where((n < 0) & (m % 2 == 1), -1.0, 1.0)

    n_max = int(m.max()) + num_terms
    fact = np.empty(n_max + 1, dtype=np.float64)
    fact[0] = 1.0
    for i in range(1, n_max + 1):
        fact[i] = fact[i - 1] * i

    result = np.zeros(np.broadcast_shapes(n.shape, x.shape), dtype=np.float64)

    half_x2 = half_x * half_x
    term_numer = half_x**m
    term_denom = fact[m]

    for k in range(num_terms):
        term_denom = fact[k] * fact[m + k]
        result += term_numer / term_denom
        term_numer *= -half_x2

    return sign * result
```

I think this fixes the agreement with scipy, but please check

astrofrog replied:
Sorry the diff is actually not that big (not all lines are changed) but it was easier to copy and paste the whole updated function

astrofrog commented on astropy/timeseries/periodograms/lombscargle/implementations/utils.py:
```diff
@@ -107,50 +394,81 @@ def trig_sum(t, h, df, N, f0=0, freq_factor=1, oversampling=5, use_fft=True, Mff
     Mfft : int
         The number of adjacent points to use in the FFT approximation.
-        Not referenced if use_fft is False.
+        Not referenced if use_fft is False  or prefer_lra is True and SciPy is available.
```
prefer_lra is not a parameter? (and scipy is not required now I think?)

astrofrog reviewed (changes requested):
Overall this looks great and seems to match the implementation described in the paper. I have a few comments below, but once these are addressed or responded to I think we can go ahead and merge this. Thanks @kkrutkowski!

kkrutkowski committed:
Add jn() tests and fix the documentation

astrofrog commented:
Before I review this again I wanted to mention that we should add an entry to the what's new in 8.0 page. Would you be able to do this? The file is in docs/whatsnew/8.0.rst

neutrinoceros commented:
@astrofrog I propose we merge this today regardless and postpone any corrections to between FF and RC

astrofrog reviewed (approved):
Let's do this - thanks @kkrutkowski!

kkrutkowski commented:
> Let's do this - thanks @kkrutkowski!

Thank you very much for pointing out the issues and sorry for takind so long to add the whatsnew entry (i still struggle a little bit with deciding what to put in there). If there are any issues with the code I'll correct them asap though.

neutrinoceros commented:
I opened https://github.com/astropy/astropy/issues/19642 to keep track of what remains to be done before rc. Thanks, all !

pllim commented:
For future reference, if this needed What's New, you should have added `whatsnew-needed` label. That would have blocked merge until it was added.

neutrinoceros commented:
> That would have blocked merge until it was added.

that is precisely why I didn't use the label actually; I really didn't want to postpone this merge to 8.1 over this.
